### PR TITLE
Use BufferedReader instead of BytesIO in `utils` subpackage

### PR DESCRIPTION
**Pull request by jairideout**
Addresses part of #6652 by using `io.BufferedReader` instead of `io.BytesIO` in the `utils` subpackage. This change was only necessary in the `astropy.utils.data.get_readable_fileobj` function.

### CODE REVIEW

File: astropy/utils/data.py [hunk 1]
```diff
@@ -197,13 +197,37 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
...
+                result = self.fileobj.read(len(b))
+                if result is None:
+                    return None
+
+                nbytes = len(result)
+                b[:nbytes] = result
+                return nbytes
+
+        stream = RawStream(fileobj)
+        fileobj = io.BufferedReader(stream)
```
Comment (saimn) [inline astropy/utils/data.py]: I don't understand why the additional `RawStream` layer is needed, instead of using directly `BufferedReader` ?

Comment (jairideout) [reply astropy/utils/data.py]: The constructor for `io.BufferedReader` requires an object that's a subclass (or at least a duck type) of `io.RawIOBase`.

In this part of the codepath for `get_readable_fileobj()`, `fileobj` is only guaranteed to have a `.read()` method, so an `io.RawIOBase` subclass is required to wrap `fileobj` in an appropriate interface for `io.BufferedReader`.

**Comment by pllim** [comment]
I am proposing a simpler approach at #8634.

**Comment by jairideout** [comment]
@pllim #8634 is indeed simpler than this PR (just wrap `BytesIO` in `BufferedReader`). I chose the implementation here to avoid reading the entire file into a `BytesIO` object, specifically to address the code comments on lines 200-204:

```
# Check if the file object supports random access, and if not,
# then wrap it in a BytesIO buffer.  It would be nicer to use a
# BufferedReader to avoid reading loading the whole file first,
# but that is not compatible with streams or urllib2.urlopen
# objects on Python 2.x.
```

Just wanted to note the differences between our solutions -- thanks!

**Comment by pllim** [comment]
@jairideout , gotcha. However, I am not sure anymore that we even need either PR, as the current code is not causing any problem AFAIK and the intent of the original comment was lost with time. :woman_shrugging:

**Comment by jairideout** [comment]
Sounds good! Feel free to close this PR if you decide it's not worth including due to the additional complexity 👍

**Comment by saimn** [comment]
I had a look but could not find a real use case for this while running the tests. Sorry for the useless work, but I think that we should close, and wait for a real need to consider using this, given the additional complexity.

**Comment by pllim** [comment]
I concur with @saimn . Thanks for the work anyway, @jairideout . It is much appreciated and we hope you continue to contribute!

**Pull request closed without merging**